File: Binary Tree - BFS/199. Binary Tree Right Side View/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        queue = deque([root])
        result = []

        while queue:
            level_size = len(queue)
            right_value = None

            for _ in range(level_size):
                node = queue.popleft()
                right_value = node.val

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(right_value)

        return result

File: Binary Tree - BFS/199. Binary Tree Right Side View/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
    if (!root) return [];

    const queue: TreeNode[] = [root];
    const result: number[] = [];

    while (queue.length) {
        const levelSize = queue.length;
        let rightValue: number | null = null;

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()!;
            rightValue = node.val;

            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }

        result.push(rightValue!);
    }

    return result;
}